rfcs/0004-categorized-attributes.md

```diff
@@ -0,0 +1,217 @@
+- Feature Name: Categorized attributes
+- Affected audience: Game Developers
+- RFC PR: [#9114](https://github.com/endless-sky/endless-sky/pull/9114)
+- Relevant Issues/RFCs: [#9025](https://github.com/endless-sky/endless-sky/pull/9025)
+
+# Summary
+
+Provides a way of programmatically accessing attributes instead of requiring the unique, string-form name of the attribute.
+
+It is a **goal** to provide programmatic access to attributes relating to runtime ship properties, such as shield amount, damage dealt by a projectile, or energy consumption.
+It is a **goal** to maintain compatibility with the old data format, including support for every existing attribute with engine support.
+
+It is a **non-goal** to provide such access to every attribute.
+It is a **non-goal** to limit the supported set of attributes.
+It is a **non-goal** to coerce the format of the string attributes to allow for generating the attribute names programmatically.
+It is a **non-goal** to maintain parity with the old data format; new attributes might receive engine support wihtout a corresponding old attribute name.
+
+# Motivation
+
+The game is rapidly expanding the number of attributes that require in-engine support. One great example can be seen in the damage types: we have shield damage, hull damage, ion damage, fuel damage, slowing damage, heat damage, etc. These damage types all have their own damage, protection, resistance etc. attributes for parity. Currently, all of these require a unique, non machine-generated string name that is used as a literal across multiple engine functions. This usage is costly to maintain due to the number of engine changes required for a new attribute, and also leaves great potential for errors due to the use of arbitrary string literals.
+
+This RFC proposes a unified system replacing the string literals used for attribute access inside the engine, offering better maintainability and, in the future, better performance as well.
+
+# Detailed Design
+
+A _categorized attribute_ is an attribute that is affected by the proposed changes, as outlined in the goals section.
+
+This RFC identifies the various components of a _categorized attribute_ as:
+- Category (when is the change made?)
+- Effect (what is changed?)
```

Comment:
Could we change "Category" to "Operation", and "Effect" to "Costs"?

```suggestion
- Operation (what is happening, what is the action we are doing or is being done?)
- Costs (what does the Operation cost in terms of energy and/or damage taken?)
```

Reply:
Right, so we would effectively support the following format:

```
"shield generation"
	shield -10
	energy 20
```

And I guess we would also support the following "backwards-compatible" format:

```
"shield generation" 10
	energy 20
```

For supporting the first format it would indeed be strange to call it costs, because the benefits are also listed. Effect is then a better name, but effect is also used for visual effects. We could consider calling it different like `result` (result of the operation), but I think effect is at least as good as result.

Reply:
I guess it also makes sense to have this one the other way around regarding signs:

```
"shield generation"
	shield 10
	energy -20
	
"shield generation" 10
	energy -20
```